src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java

```diff
@@ -512,108 +548,69 @@ private void alignSlotData()
-                    // Out of slab memory
-                    processInvalidRequest(503, "Service Unavailable");
+                    processInvalidRequest(400, "Bad Request");
                 }
-                else
+                if (length >= maximumHeadersSize)
                 {
-                    slotOffset = 0;
-                    MutableDirectBuffer buffer = slab.buffer(slotIndex);
-                    buffer.putBytes(0, payload, offset, length);
-                    slotPosition = length;
-                    int firstSpace = limitOfBytes(buffer, slotOffset, slotPosition, SPACE);
-                    if (firstSpace > MAXIMUM_METHOD_BYTES)
+                    int firstCRLF = limitOfBytes(payload, offset, limit, CRLF_BYTES);
+                    if (firstCRLF == -1 || firstCRLF > maximumHeadersSize)
                     {
-                        processInvalidRequest(400, "Bad Request");
+                        processInvalidRequest(414, "Request URI too long");
                     }
-                    if (window == 0)
+                    else
                     {
-                        // Increase source window to ensure we can receive the largest possible request headers
-                        ensureSourceWindow(maximumHeadersSize - length);
-                        if (window < 2)
-                        {
-                            int firstCRLF = limitOfBytes(buffer, slotOffset, slotPosition, CRLF_BYTES);
-                            if (firstCRLF == -1 || firstCRLF > maximumHeadersSize)
-                            {
-                                processInvalidRequest(414, "Request URI too long");
-                            }
-                            else
-                            {
-                                processInvalidRequest(431, "Request Header Fields Too Large");
-                            }
-                        }
+                        processInvalidRequest(431, "Request Header Fields Too Large");
                     }
                 }
             }
             else
             {
                 decodeCompleteHttpBegin(payload, offset, endOfHeadersAt - offset);
-                result = slotOffset = endOfHeadersAt;
+                result = endOfHeadersAt;
             }
             return result;
-        }
+        };
+
+        private final DecoderState decodeBeforeHttpBegin = (payload, offset, limit) ->
+        {
+            int length = limit - offset;
+            int result = offset;
+            if (payload.getByte(offset) == '\r')
+            {
+                if (length > 1 && payload.getByte(offset+1) == '\n')
+                {
+                    // RFC 3270 3.5.  Message Parsing Robustness: skip empty line (CRLF) before request-line
+                    result = offset + 2;
+                }
+            }
+            else
+            {
+                decoderState = decodeHttpBegin;
+            }
+            return result;
+        };
```

Comment:
Which you prefer me to switch all the decoder states back to normal methods and reference them all as lambdas as before? My intention was to move all of them to member variables for efficiency and clarity, but unfortunately I got burnt by the need to make a forward reference so at least one of them had to remain a normal method.